File: shared/src/main/java/ru/privatenull/pnauth/kernel/service/ServiceRegistry.java
package ru.privatenull.pnauth.kernel.service;
import java.util.List;
import java.util.Optional;
public interface ServiceRegistry {
    <T> ServiceRegistration register(ServiceKey<T> key, String ownerId, int priority, T service);
    <T> Optional<T> find(ServiceKey<T> key);
    <T> List<T> findAll(ServiceKey<T> key);
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/kernel/service/ServiceRegistry.kt
package ru.privatenull.pnauth.kernel.service

import java.util.Optional

interface ServiceRegistry {
    fun <T> register(key: ServiceKey<T>, ownerId: String, priority: Int, service: T): ServiceRegistration
    fun <T> find(key: ServiceKey<T>): Optional<T>
    fun <T> findAll(key: ServiceKey<T>): List<T>
}
